### DISCUSSION TITLE
What's the correct height of a bin + lip?

### DISCUSSION THREAD

**Discussion by GCheung55** [kennetek/gridfinity-rebuilt-openscad#197] [Q&A] [answered]
Assume that the bin is 1x1x2. So it.s 41.5mm x 41x5mm x 14mm.

`gridfinity-rebuilt-openscad` says the lip should be 3.8, making the bin height 17.8mm.

[Specs](https://gridfinity.xyz/specification/) shows the lip should be 4.4, making the bin height 18.4mm.

What's the correct bin height?

**Comment by EmperorArthur** [comment]
Will edit this when I am at my PC, and can look at the code. 

The spec assumes the lip ends at a sharp point.  However, we radius the top of the lip, lowering the height slightly.

I am the one who wrote the current profile code describing the lip based on that spec, and believe there is a picture included in the images folder which maps the lip to measurements in standard.scad

**Reply by GCheung55** [reply]
I see. I didn't find the picture - my search-fu is lacking.

Based on the comment in https://github.com/kennetek/gridfinity-rebuilt-openscad/blob/main/gridfinity-rebuilt-bins.scad#L16, the height of a 1x1x2 bin should be 17.8mm then. However, with the `h_lip` currently being 3.548, the output I'm seeing is 17.515mm.

Should the `h_lip` actually be 4.4? That would output a height of 17.8mm. Otherwise the comment in gridfinity-rebuilt-bins.scad is misleading.

**Reply by EmperorArthur** [reply]
Back at my PC so, two points.

### h_lip Is Not Arbitrary, And Cannot Be Changed

While I did not come up with that number, I was able to re-calculate it, and experimented with calculating it on the fly.

The stacking lip from the specification is 4.4mm high (stacking_lip_height).  Using the tangent tangent radius algorithm to determine the center of a 0.6mm (r_f1) radius, then adding that radius back gives 3.548mm.

The code actually draws out a sharp top lip, then cuts the top off and replaces it with a radius.  Unfortunately, since h_lip is pre-calculated, you can't change it or r_f1 without causing issues with the model.

### Not Seeing The Problem

How are you measuring that?  Because, I am not seeing it.

The attached image shows 12.174mm.

2 units is 14mm.  Remove 5mm for the base (h_base), and you get 9mm.  Which is what is passed to the function to create the walls.  That's then going to add on the 3.458mm, which gives 12.458mm.  Except I'm measuring at the base of the radius.  So would expect to see 0.6mm less.  Which would be 11.858mm.

 Close enough for quick point measurements with OpenSCAD.  Especially since I haven't bothered to verify in the code that the lower point really is 5mm high, and I wouldn't be surprised if it isn't, as the code which generates it does so in a way that can take a bit to wrap your head around.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/e3700027-c455-45e6-8bc2-bd1e426a3caf)

**Reply by GCheung55** [reply]
I see. I got the dimensions by exporting an STL and looking at the dimensions in OrcaSlicer.
<img width="723" alt="image" src="https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/46033/053b8ab0-9c10-4412-bfa4-224532cd25fa">

**Reply by GCheung55** [reply]
If `h_lip` is changed to 4.4, the dimensions I see in OrcaSlicer is:
<img width="981" alt="image" src="https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/46033/bc9d379a-7dcf-4f66-806b-5582ef5bdfe9">

Can you help me understand what issues happen to the model?

**Reply by EmperorArthur** [reply]
### Congratulations

Congratulations, you found a bug.  Just not the one you think.  Also not one I want to look at fixing until I can get some of my current work in progress merged. 

### h_lip

Take a close look at the top of the lip when setting `h_lip=4.4;`  You will see the top is no longer rounded, and is instead flat.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/bf02ed11-fe54-4168-95e5-411b817179a3)

### Wall Profile

I ran the following to generate the profile used for the wall:
```
$fa = 8;
$fs = 0.25;
profile_wall(9);
```

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/1355f261-54b4-4287-a2f0-b96d37ee013a)

I previously gave the number of 11.858mm, and I'd say 0.1mm off is pretty close.  Especially given how OpenSCAD does circles.

### PrusaSlicer Measurements

Commenting out the base, I see 12.55mm.  Which is within tolerance of the 12.458mm I calculated earlier.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/b6d50f4b-f5d0-4b90-a70d-6a7f344aabee)

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/ffb3711b-587a-4b66-adae-0d52c3d94193)

I then loaded just the base and got these numbers.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/fc16164e-b6ba-420e-9ef9-150e2bcef105)

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/ce11ce1a-57c5-4c22-831f-139338bce58f)


### FreeCAD Measurements

I have loaded the model into FreeCAD, converted it to a wireframe, and took some measurements.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/2c9f89f7-23a6-4352-8548-56c50df924d8)

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/b6fa2181-4dc9-46f0-8c9b-6a210949ef75)

4.788+12.174+0.6 == 17.562

### The Bug

Take another look at the FreeCAD Wireframe.  Does something look odd to you?

Here's a close up, with the error measured.

![image](https://github.com/kennetek/gridfinity-rebuilt-openscad/assets/1515637/495cfd12-4d80-4dd3-bf99-12e24e0c7b0b)

That's the source of the problem, but it isn't jumping out at me in the code, and the part which creates the base takes a significantly different approach than the wall and my PR for the baseplate.

**Reply by GCheung55** [reply]
> Congratulations, you found a bug. Just not the one you think. Also not one I want to look at fixing until I can get some of my current work in progress merged.

Yay, glad I could help.

If I'm understanding you correctly, the bug is that there's 0.367mm that is causing the height to be wrong? Any suggestions how I could mitigate the issue until you have time?

**Reply by GCheung55** [reply]
> Take a close look at the top of the lip when setting `h_lip=4.4;` You will see the top is no longer rounded, and is instead flat.

Is there something wrong with having a flat top instead of a radius? Just trying to understand if/why that's an issue.

I may have misunderstood, and the bug isn't actually impacting the height of the bin. If a 1x1x2 bin with the lip is supposed to be 17.5515mm tall, then perhaps the [code comment](https://github.com/kennetek/gridfinity-rebuilt-openscad/blob/main/gridfinity-rebuilt-bins.scad#L16) needs to be updated, since it's saying that the height of the bin is supposed to be 17.8mm tall. If that's the case, I can submit a PR to make the correction.

**Reply by EmperorArthur** [reply, marked as answer]
> If I'm understanding you correctly, the bug is that there's a 0.367mm reduction that is causing the height to be wrong?

Looking into the issue some more, it does appear to be an issue with something in how the radius is applied to the `profile_wall` code.  Verified by first fixing the issue with the base.

> Any suggestions how I could mitigate the issue until you have time?

Ignore it.  I've done some basic testing by removing the chamfer completely and the lip was the expected 4.4mm height.  The only consequence is the chamfer being slightly lower than it should be.  However, as the first inner transition is not affected, bins will stack exactly as if the lip came to a sharp point.

Remember, the spec notes a 0.5mm tolerance.  So, while this should be fixed, it' not that big of a deal.
